src/source.rs

```diff
@@ -0,0 +1,115 @@
+
+    let temperature = root
+        .get("temperature")
+        .and_then(|v| v.get("current"))
+        .and_then(|v| v.as_u64())
+        .and_then(|v| v.try_into().ok());
+
+    Ok(temperature)
+}
+
+/// Get the temperature from a plain-text file (typically a sysfs path). This
+/// function only fails if the file cannot be read or if the whitespace-trimmed
+/// contents cannot be parsed as an integer. If the reported temperature exceeds
+/// the bounds of a `u8`, then `Ok(None)` is returned.
+fn parse_file_source<T: AsRef<Path>>(path: T) -> Result<Option<u8>> {
+    let contents = fs::read_to_string(path.as_ref())
+        .context(IoError { path: path.as_ref().to_owned() })?;
+    let trimmed = contents.trim();
+
+    // The file should be in milli-degrees Celsius
+    let temperature = trimmed
+        .parse::<u32>()
+        .context(SensorValueParseError { value: trimmed.to_owned() })?
+        .checked_div(1000)
+        .and_then(|t| t.try_into().ok());
+
+    Ok(temperature)
+}
+
+/// Get the temperatures for the given list of sensors from IPMI. This performs
+/// one ipmitool query to reduce the IPMI round trips, which may be slow. This
+/// function only fails if the IPMI sensor query fails. If a sensor's unit is
+/// not degrees Celsius or if the value exceeds the bounds of a `u8`, then the
+/// reported value of that sensor will be `None`.
+fn parse_ipmi_sources<'a, T: AsRef<str>>(ipmi: &mut Ipmi, sensors: &'a [T])
+    -> Result<HashMap<&'a str, Option<u8>>>
+{
+    let ipmi_readings = ipmi.get_sensor_readings(sensors)
+        .context(IpmiError)?
+        .into_iter()
+        .collect::<Result<Vec<SensorReading>, _>>()
+        .context(IpmiError)?;
+
+    let result = ipmi_readings
+        .into_iter()
+        .map(|r| {
+            if r.units == "degrees C" {
+                r.value.parse::<u8>().ok()
+            } else {
+                None
+            }
+        })
+        .zip(sensors.iter())
+        .map(|(temp, name)| (name.as_ref(), temp))
+        .collect();
+
+    Ok(result)
+}
+
+/// Get temperature readings for the given sources. The returned values are in
+/// the same order as given.
+pub fn get_source_readings(ipmi: &mut Ipmi, sources: &[Source]) -> Result<Vec<Option<u8>>> {
+    // Get IPMI sensor readings in one go for better performance.
+    let ipmi_sensors = sources.iter()
+        .filter_map(|s| {
+            match s {
+                Source::Ipmi { sensor } => Some(sensor),
+                _ => None,
+            }
+        })
+        .collect::<Vec<_>>();
+
+    let ipmi_results = parse_ipmi_sources(ipmi, &ipmi_sensors)?;
+
+    sources.iter()
+        .map(|s| {
+            match s {
+                Source::Ipmi { sensor } => Ok(ipmi_results[sensor.as_str()]),
+                Source::File { path } => parse_file_source(path),
+                Source::Smart { block_dev } => parse_smart_source(block_dev),
```

Comment:
I have to imagine the "hangs" I'm seeing are due to waiting for every block device to return a result 

If I'm reading the changes in https://github.com/chenxiaolong/ipmi-fan-control/commit/156d3ae342ec832c50783425af70f4ea10a40919 there is a single loop that  handles all zones. 

If a loop is slow, we'll miss a tick. 

Am I assessing this correctly? 

If so, a far future update may break each zone out into it's own loop so that something slow like collecting smart data wouldn't hold up the CPU zone. 

Alternatively one could optimize the loop by having the source readings happen asynchronously, writing to an in memory store. 

Please note, this is _not_ a request for change, just passing back notes from running it for 30 minutes. [The Perl script I modified](https://github.com/khorton/nas_fan_control/) suffered from the same (and far worse) problems

Reply:
> I'm noticing some "hangs" during zone switching. Note the 5 second gap between logs. I have a guess as to what is causing this in a comment below.

> If I'm reading the changes in 156d3ae correctly there is a single loop that handles all zones.
>
> If a loop is slow, we'll miss a tick.
>
> Am I assessing this correctly?

Yep, there's a single loop, though it's a very "dumb" loop. If querying the sensors takes a long time, the next tick is delayed instead of being skipped.

> If so, a far future update may break each zone out into it's own loop so that something slow like collecting smart data wouldn't hold up the CPU zone.

I have an `async` branch I'm testing locally that implements this :D My motivation was actually for handling `^C`/`SIGINT`/`SIGTERM` more easily (especially on Windows). I expect to have a PR up for this in the next day or so.

Reply:
I may also add support for querying all of the source in parallel. It should be pretty easy to do once the async code is done.